llvm/test/CodeGen/AMDGPU/atomic_optimizations_global_pointer.ll

```diff
@@ -1725,106 +2344,1350 @@ entry:
+; GFX1264_DPP-NEXT:    s_mov_b64 s[4:5], s[6:7]
+; GFX1264_DPP-NEXT:    v_writelane_b32 v1, s10, 48
+; GFX1264_DPP-NEXT:    v_writelane_b32 v2, s11, 48
+; GFX1264_DPP-NEXT:    s_mov_b64 exec, s[8:9]
+; GFX1264_DPP-NEXT:    s_mov_b32 s6, -1
+; GFX1264_DPP-NEXT:    s_mov_b64 s[8:9], exec
+; GFX1264_DPP-NEXT:    ; implicit-def: $vgpr8_vgpr9
+; GFX1264_DPP-NEXT:    v_cmpx_eq_u32_e32 0, v0
+; GFX1264_DPP-NEXT:    s_cbranch_execz .LBB5_2
+; GFX1264_DPP-NEXT:  ; %bb.1:
+; GFX1264_DPP-NEXT:    v_mov_b32_e32 v9, s5
+; GFX1264_DPP-NEXT:    v_mov_b32_e32 v8, s4
+; GFX1264_DPP-NEXT:    s_mov_b32 s7, 0x31016000
+; GFX1264_DPP-NEXT:    s_wait_kmcnt 0x0
+; GFX1264_DPP-NEXT:    s_mov_b32 s4, s2
+; GFX1264_DPP-NEXT:    s_mov_b32 s5, s3
+; GFX1264_DPP-NEXT:    buffer_atomic_add_u64 v[8:9], off, s[4:7], null th:TH_ATOMIC_RETURN
+; GFX1264_DPP-NEXT:    s_wait_loadcnt 0x0
+; GFX1264_DPP-NEXT:    global_inv scope:SCOPE_DEV
+; GFX1264_DPP-NEXT:  .LBB5_2:
+; GFX1264_DPP-NEXT:    s_or_b64 exec, exec, s[8:9]
+; GFX1264_DPP-NEXT:    s_wait_kmcnt 0x0
+; GFX1264_DPP-NEXT:    v_readfirstlane_b32 s2, v8
+; GFX1264_DPP-NEXT:    v_mov_b32_e32 v10, v1
+; GFX1264_DPP-NEXT:    v_mov_b32_e32 v11, v2
+; GFX1264_DPP-NEXT:    v_readfirstlane_b32 s3, v9
+; GFX1264_DPP-NEXT:    s_delay_alu instid0(VALU_DEP_3) | instskip(SKIP_1) | instid1(VALU_DEP_2)
+; GFX1264_DPP-NEXT:    v_add_co_u32 v8, vcc, s2, v10
+; GFX1264_DPP-NEXT:    s_mov_b32 s2, s6
+; GFX1264_DPP-NEXT:    v_add_co_ci_u32_e32 v9, vcc, s3, v11, vcc
+; GFX1264_DPP-NEXT:    s_mov_b32 s3, 0x31016000
+; GFX1264_DPP-NEXT:    buffer_store_b64 v[8:9], off, s[0:3], null
+; GFX1264_DPP-NEXT:    s_nop 0
+; GFX1264_DPP-NEXT:    s_sendmsg sendmsg(MSG_DEALLOC_VGPRS)
+; GFX1264_DPP-NEXT:    s_endpgm
+;
+; GFX1232_DPP-LABEL: add_i64_varying:
+; GFX1232_DPP:       ; %bb.0: ; %entry
+; GFX1232_DPP-NEXT:    v_mov_b32_e32 v8, v0
+; GFX1232_DPP-NEXT:    s_or_saveexec_b32 s2, -1
+; GFX1232_DPP-NEXT:    v_mov_b32_e32 v1, 0
+; GFX1232_DPP-NEXT:    v_mov_b32_e32 v2, 0
+; GFX1232_DPP-NEXT:    s_mov_b32 exec_lo, s2
+; GFX1232_DPP-NEXT:    v_mov_b32_e32 v9, 0
+; GFX1232_DPP-NEXT:    v_mov_b32_e32 v3, v8
+; GFX1232_DPP-NEXT:    s_delay_alu instid0(VALU_DEP_2)
+; GFX1232_DPP-NEXT:    v_mov_b32_e32 v4, v9
+; GFX1232_DPP-NEXT:    s_not_b32 exec_lo, exec_lo
+; GFX1232_DPP-NEXT:    v_mov_b32_e32 v3, 0
+; GFX1232_DPP-NEXT:    v_mov_b32_e32 v4, 0
+; GFX1232_DPP-NEXT:    s_not_b32 exec_lo, exec_lo
+; GFX1232_DPP-NEXT:    s_or_saveexec_b32 s2, -1
+; GFX1232_DPP-NEXT:    v_dual_mov_b32 v5, v2 :: v_dual_mov_b32 v6, v2
+; GFX1232_DPP-NEXT:    v_add_co_u32_e64_dpp v3, vcc_lo, v3, v3 row_shr:1 row_mask:0xf bank_mask:0xf bound_ctrl:1
+; GFX1232_DPP-NEXT:    v_mov_b32_e32 v7, v2
+; GFX1232_DPP-NEXT:    s_delay_alu instid0(VALU_DEP_3) | instskip(NEXT) | instid1(VALU_DEP_1)
+; GFX1232_DPP-NEXT:    v_mov_b32_dpp v5, v4 row_shr:1 row_mask:0xf bank_mask:0xf
+; GFX1232_DPP-NEXT:    v_add_co_ci_u32_e32 v4, vcc_lo, v4, v5, vcc_lo
```

Comment:
As a follow up we should try to improve this DPP code. The docs say that v_add_co_ci_u32_e32 has a DPP form. I don't know why we're not using it here.